### ISSUE TITLE
Prompt command hangs fish

### ISSUE THREAD

**Issue by nyarly** [fish-shell/fish-shell#5817]
I have a function that's part of my config, a variant of `__fish_git_prompt.fish` that calls this pipeline:
```
timeout 0.5 git diff --name-status | cut -c 1-2;
```
The whole section looks like:
```
if type -q timeout
    set changedFiles (timeout 0.5 git diff --name-status | cut -c 1-2; or echo 0)
    set stagedFiles (timeout 0.5 git diff --staged --name-status | cut -c 1-2; or echo 0)
    set untrackedfiles (count (timeout 0.5 git ls-files --others --exclude-standard; or echo))
else 
```

Running fish with `-d3`, the output ends with it trying to execute `timeout 0.5 git diff --name-status | cut -c 1-2`

strace attached to the fish process shows a loop wait4ing for children and getting ECHILD.

I think this is related to the pgrp problem in #5765, but a) it's completely consistent and b) it doesn't happen at the prompt, only when run within the `fish_prompt` function.

I'm running fish 3.0.2.
The issue occurs in xterm and alacritty, with and without tmux.
This configuration was working fine in 2.7

**Comment by manfredlotz** [comment]
Same here: Fedora 29, fish 3.0.2 running in an xfce4-terminal

**Comment by mqudsi** [comment]
Are you able to try this with the code from `master`?

**Comment by manfredlotz** [comment]
I cloned fish-shell, built it, installed into /usr/local. To be on the safe side I removed the system package.

No hanging situation any longer. So, it seems it is solved in master

**Comment by ridiculousfish** [comment]
Cool, good questions to answer are:

1. which commit fixed this (`git bisect`)
2. was this in 3.0.0 and was it in 2.7.x

that will help us decide whether to take for 3.0.3

**Comment by mqudsi** [comment]
Yup. @Ridiculousfish 

> This configuration was working fine in 2.7

**Comment by faho** [comment]
Okay, to reproduce, I have to incorporate the command substitution - `function fish_prompt; set changedFiles (timeout 0.5 git diff --name-status | cut -c 1-2; or echo 0); echo dollar; end` shows the issue.

Bisect reveals the fix to be 9a4153f5e2f0a0ea74f395248f5dbd40b55e8dc8. Which... isn't good news, since that's a large rearchitecturing that required a bunch of fixes on top, so I don't think we want to backport it.

Which means I'm closing this as fixed in 3.**1**, not 3.0.3.
